apps/labs/posts/python-wheels-from-tags-to-variants.md

```diff
@@ -0,0 +1,1146 @@
+requires = ["provider-variant-x86-64 >=0.0.1,<1"]
+# use only on x86_64 machines
+enable-if = "platform_machine == 'x86_64' or platform_machine == 'AMD64'"
+plugin-api = "provider_variant_x86_64.plugin:X8664Plugin"`
+```
+
+<figcaption>Listing 3. Example `pyproject.toml` with sort order defined</figcaption>
+</figure>
+
+However, no reason to stop at namespace. After all, a specific package may
+want to point out, say, that their AES-NI variant is more performant than
+a fallback implemented using the instruction sets provided by x86-64-v3.
+It might also be desirable reorder value; while
+I immediately can't think of a specific use case for that, let's have the option
+for symmetry anyway. So while the plugins specify the initial order of features
+and their values, packages are allowed to override it.
+
+On top of that, user configuration can be applied. After all, there is still
+a valid use case for the user to override the sort order defined by plugins
+and packages. And since it's no longer obligatory, having that is not a problem.
+
+## The null variant
+
+Let's consider that you are distributing a package with three CUDA variants,
+and a CPU variant.
+
+<figure>
+
+```rust
+torch-2.8.0-cp313-cp313-manylinux_2_28_x86_64-cu129.whl
+torch-2.8.0-cp313-cp313-manylinux_2_28_x86_64-cu128.whl
+torch-2.8.0-cp313-cp313-manylinux_2_28_x86_64-cu126.whl
+torch-2.8.0-cp313-cp313-manylinux_2_28_x86_64.whl        // CPU-only
+```
+
+<figcaption>Listing 4. Example wheel variant filenames, with a fallback regular wheel, in order of preference</figcaption>
+</figure>
+
+Such a setup provides for a graceful fallback. Depending on your exact CUDA version,
+the top variants can be filtered out, and the lower variants will be used instead.
+If you don't have a compatible CUDA runtime at all, the fallback CPU wheel
+is used. So far, so good.
+
+However, consider a system with an older package manager version that does not
+support variants. Independently of the presence of CUDA runtime, all variant
+wheels are ignored there, and the CPU-only fallback is installed. However,
+this is not the most optimal solution. Prior to introducing variants,
+the published wheels may have featured both CUDA and CPU support (even if for a single
+CUDA version).
+
+This is where the null variant comes in. Consider the following instead.
+
+<figure>
+
+```rust
+torch-2.8.0-cp313-cp313-manylinux_2_28_x86_64-cu129.whl
+torch-2.8.0-cp313-cp313-manylinux_2_28_x86_64-cu128.whl
+torch-2.8.0-cp313-cp313-manylinux_2_28_x86_64-cu126.whl
+torch-2.8.0-cp313-cp313-manylinux_2_28_x86_64-00000000.whl  // CPU-only
+torch-2.8.0-cp313-cp313-manylinux_2_28_x86_64.whl           // CUDA 12.6 + CPU
+```
+
+<figcaption>Listing 5. Example wheel variant filenames, with a null variant and a fallback regular wheel, in order of preference</figcaption>
+</figure>
+
+What we added here is a null variant, with label `00000000`. Since it has
+no properties, it is always supported; that is, as long as variants
```

Comment:
Where is the lack of properties for that variant defined? Surely it doesn't come from a magic string `00000000` (and in the unlikely case that this is so, what about `0000000` or `000000000`?)